File: C++/IsGraphBipartite.cpp
//Ian Hays
//02-14-2021
//https://leetcode.com/problems/is-graph-bipartite/
//SC: O(V*E) TC: O(V*E)

class Solution {
public:
    bool isBipartite(vector<vector<int>>& graph) {
        queue<array<int,2>> q;
        vector<int> visited(graph.size(),0);
        for(int i = 0; i < visited.size(); i++){
            if(visited[i]) continue;
            q.push({i,1});
            while(!q.empty()){
                int size = q.size();
                for(int j = 0; j < size; j++){
                    auto [node,set] = q.front();
                    q.pop();
                    set *= -1;
                    for(int k = 0; k < graph[node].size(); k++){
                        int neighbor = graph[node][k];
                        if(!visited[neighbor]){
                            visited[neighbor] = set; 
                            q.push({neighbor,set});
                        } else {
                            if(visited[neighbor] != set) return false;
                        }
                    }
                }
            }
        }
        return true;
    }
};

File: C#/IsGraphBipartite.cs
//Ian Hays
//02-14-2021
//https://leetcode.com/problems/is-graph-bipartite/
//SC: O(V*E) TC: O(V*E)

public class Solution {
    public bool IsBipartite(int[][] graph) {
        Queue<int[]> queue = new Queue<int[]>();
        int[] visited = new int[graph.Length];
        for(int i = 0; i < visited.Length; i++){
            if(visited[i] != 0) continue;
            queue.Enqueue(new int[2]{i,1});
            while(queue.Count != 0){
                int size = queue.Count;
                for(int j = 0; j < size; j++){
                    int node = queue.Peek()[0];
                    int set = queue.Peek()[1]*-1;
                    queue.Dequeue();
                    for(int k = 0; k < graph[node].Length; k++){
                        int neighbor = graph[node][k];
                        if(visited[neighbor] == 0){
                            visited[neighbor] = set; 
                            queue.Enqueue(new int[2]{neighbor,set});
                        } else {
                            if(visited[neighbor] != set) return false;
                        }
                    }
                }
            }
        }
        return true;
    }
}
